data-connection/src/main/java/io/micronaut/data/connection/annotation/ConnClientInfoAttr.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2017-2024 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micronaut.data.connection.annotation;
+
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * An annotation used to set client info for the connection.
+ *
+ * @author radovanradic
+ * @since 4.11
+ */
+@Target({ElementType.ANNOTATION_TYPE, ElementType.METHOD, ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Repeatable(ConnClientInfo.class)
+@Connectable
+public @interface ConnClientInfoAttr {
```

Comment:
I would rename this to just `@ClientInfo` (the word `connection` is already in the package) and make the repeatable annotation an annotation named `List` that is an inner type. See https://github.com/sergiy-naumovych/Professional-Java-for-Web-Applications/blob/master/Chapter%2021/Eclipse/Spring-JPA/source/production/java/com/wrox/site/validation/NotBlank.java for an example

Reply:
Then repository looks super awkward 
```
@JdbcRepository(dialect = Dialect.ORACLE)
@ClientInfo.List
public interface OracleXEAuthorRepository extends AuthorRepository {
    @Override
    @Join(value = "books", type = Join.Type.LEFT_FETCH)
    @ClientInfo(name = "OCSID.ACTION", value = "QueryAuthorByName")
    Author queryByName(String name);
}
```
Please suggest something new, I have no more ideas.

Reply:
why is the annotation needed on the type?

Reply:
Because from type we get module and don't have to annotate each method. Then module will be repository class name and action will be method name. But, it can be added on the method as well, for example if users want to name module/action differently.

Reply:
Yes, but if want to allow fallback for module attribute (to the repository class name) then we do this 
```
@JdbcRepository(dialect = Dialect.ORACLE)
@ClientInfo
public interface OracleXEAuthorRepository extends AuthorRepository {
    @Override
    @Join(value = "books", type = Join.Type.LEFT_FETCH)
    @ClientInfo.Attribute(name = "OCSID.ACTION", value = "QueryAuthorByName")
    Author queryByName(String name);
}
```
and on methods we can override module or action attributes. That's how it looks like now and I think this is fine.

Reply:
I don't think `@ClientInfo` without anything makes sense. I would expect exactly what Graeme suggested; otherwise, it needs to be some custom annotation that will transform the method/package into custom client info.

Another option would have some placeholders like `MD.METHOD_NAME` which we will replace at runtime.

Reply:
Changed now since we can use metadata from `SqlPreparedQuery` when executing calls and just populate defaults if not provided in the annotations. And will be applied to all Oracle jdbc calls if `datasources.<datsource-name>.enable-oracle-client-info=true`